File: android/personstate/src/test/kotlin/com/farosos/personstate/FakeScheduler.kt
package com.farosos.personstate

/**
 * `Scheduler` de prueba: el tiempo solo avanza cuando el test llama a
 * `advance(by:)`, así los tests corren instantáneo sin depender de esperar
 * minutos reales para el timer de gracia o el timeout.
 */
class FakeScheduler : Scheduler {
    private class Entry(val id: Int, val fireTime: Double, val action: () -> Unit) : SchedulerToken

    private val entries = mutableListOf<Entry>()
    private var now = 0.0
    private var nextId = 0

    override fun schedule(afterSeconds: Double, action: () -> Unit): SchedulerToken {
        nextId += 1
        val entry = Entry(nextId, now + afterSeconds, action)
        entries.add(entry)
        return entry
    }

    override fun cancel(token: SchedulerToken) {
        val entry = token as? Entry ?: return
        entries.removeAll { it.id == entry.id }
    }

    /**
     * Avanza el reloj y dispara, en orden, cualquier acción programada cuyo
     * tiempo ya se cumplió.
     */
    fun advance(by: Double) {
        now += by
        while (true) {
            val due = entries.filter { it.fireTime <= now }.minByOrNull { it.fireTime } ?: break
            entries.removeAll { it.id == due.id }
            due.action()
        }
    }
}

File: ios/Tests/PersonStateMachineTests/FakeScheduler.swift
import Foundation
@testable import PersonStateMachine

/// `Scheduler` de prueba: el tiempo solo avanza cuando el test llama a
/// `advance(by:)`, así los tests corren instantáneo sin depender de esperar
/// minutos reales para el timer de gracia o el timeout.
final class FakeScheduler: Scheduler {
    private final class Entry: SchedulerToken {
        let id: Int
        let fireTime: TimeInterval
        let action: () -> Void
        init(id: Int, fireTime: TimeInterval, action: @escaping () -> Void) {
            self.id = id
            self.fireTime = fireTime
            self.action = action
        }
    }

    private var entries: [Entry] = []
    private var now: TimeInterval = 0
    private var nextId = 0

    func schedule(after seconds: TimeInterval, _ action: @escaping () -> Void) -> SchedulerToken {
        nextId += 1
        let entry = Entry(id: nextId, fireTime: now + seconds, action: action)
        entries.append(entry)
        return entry
    }

    func cancel(_ token: SchedulerToken) {
        guard let entry = token as? Entry else { return }
        entries.removeAll { $0.id == entry.id }
    }

    /// Avanza el reloj y dispara, en orden, cualquier acción programada cuyo
    /// tiempo ya se cumplió.
    func advance(by seconds: TimeInterval) {
        now += seconds
        while let due = entries.filter({ $0.fireTime <= now }).min(by: { $0.fireTime < $1.fireTime }) {
            entries.removeAll { $0.id == due.id }
            due.action()
        }
    }
}
